src/CircuitBreakerPolicy.ts

```diff
@@ -0,0 +1,185 @@
+    return this.innerState.value;
+  }
+
+  /**
+   * Gets the last reason the circuit breaker failed.
+   */
+  public get lastFailure() {
+    return this.innerLastFailure;
+  }
+
+  constructor(private readonly options: ICircuitBreakerOptions<R>) {}
+
+  /**
+   * Manually holds open the circuit breaker.
+   * @returns A handle that keeps the breaker open until `.dispose()` is called.
+   */
+  public isolate() {
+    if (this.innerState.value !== CircuitState.Isolated) {
+      this.innerState = { value: CircuitState.Isolated, counters: 0 };
+      this.breakEmitter.emit({ isolated: true });
+    }
+
+    this.innerState.counters++;
+
+    let disposed = false;
+    return {
+      dispose: () => {
+        if (disposed) {
+          return;
+        }
+
+        disposed = true;
+        if (this.innerState.value === CircuitState.Isolated && !--this.innerState.counters) {
+          this.innerState = { value: CircuitState.Closed };
+          this.resetEmitter.emit();
+        }
+      },
+    };
+  }
+
+  /**
+   * Executes the given function.
+   * @param fn -- Function to run
+   * @throws a {@link BrokenCircuitError} if the circuit is open
+   * @throws a {@link IsolatedCircuitError} if the circuit is held
+   * open via {@link CircuitBreakerPolicy.isolate}
+   * @returns a Promise that resolves or rejects with the function results.
+   */
+  public async execute<T extends R>(fn: () => Promise<T> | T): Promise<T> {
+    const state = this.innerState;
+    switch (state.value) {
+      case CircuitState.Closed:
+        const result = await execute(this.options, fn);
+        if ('success' in result) {
+          this.options.breaker.success(state.value);
+        } else {
+          this.innerLastFailure = result;
+          if (this.options.breaker.failure(state.value)) {
+            this.open(result);
+          }
+        }
+
+        return returnOrThrow(result);
+
+      case CircuitState.HalfOpen:
+        await state.test.catch(() => undefined);
+        return this.execute(fn);
+
+      case CircuitState.Open:
+        if (Date.now() - state.openedAt < this.options.halfOpenAfter) {
+          throw new BrokenCircuitError();
+        }
+        const test = this.halfOpen(fn);
+        this.innerState = { value: CircuitState.HalfOpen, test };
+        return test;
+
+      case CircuitState.Isolated:
+        throw new IsolatedCircuitError();
+
+      default:
+        throw new Error(`Unexpected circuit state ${state}`);
+    }
+  }
+
+  private async halfOpen<T extends R>(fn: () => Promise<T> | T): Promise<T> {
+    try {
+      const result = await execute(this.options, fn);
+      if ('success' in result) {
+        this.options.breaker.success(CircuitState.HalfOpen);
+        this.close();
+      } else {
+        this.innerLastFailure = result;
+        this.options.breaker.failure(CircuitState.HalfOpen);
+        this.open(result);
+      }
+
+      return returnOrThrow(result);
+    } catch (err) {
```

Comment:
Use `finally` instead, saves a line :)

Reply:
I don't want `close()` to be called if we get a normal return

```
try {
  return;
} finally {
  console.log('this still runs!');
}
```